awspub/notification.py

```diff
@@ -0,0 +1,98 @@
+"""
+Methods used to handle notifications for AWS using SNS
+"""
+
+from typing import Optional, Dict, Any
+from mypy_boto3_sns.client import SNSClient
+from mypy_boto3_ec2.client import EC2Client
+from awspub.exceptions import AWSNotificationException
+import boto3
+import logging
+
+logger = logging.getLogger(__name__)
+email_template_path = "./awspub/email_template.txt"
+
+
+class SNSNotification(object):
+    """
+    A data object that contains validation logic and
+    structuring rules for SNS notification JSON
+    """
+
+    def __init__(self, conf: Dict[str, Any], image_name: str, image_id: str, region_name: str):
+        """
+        Construct a message and verify that it is valid
+        """
+        self._conf = conf
+        self._image_name = image_name
+        self._image_id = image_id
+        self._region_name = region_name
+
+    def _get_topic(self, client: SNSClient, topic: str) -> Optional[str]:
+        """
+        Check if topic already exist
+        """
+        resp = client.list_topics()
+        for topic_info in resp["Topics"]:
+            # Matching topic name at the end of ARN
+            if topic in topic_info["TopicArn"]:
+                return topic_info["TopicArn"]
+        return None
+
+    def _create_subject(self) -> str:
+        """
+        Create subject line for SNS notification
+
+        :return: string of subject line
+        :rtype: str
+        """
+
+        return f"A New AWS image is available {self._region_name}"
+
+    def _get_message(self) -> str:
+        """
+        Load message template to send with SNS
+
+        :return: string block of email template
+        :rtype: str
+        """
+
+        message = f"""
+Hello,
+
+New AWS image is available in {self._region_name}.
+
+{self._image_name}
+{self._image_id}
+
+Thank you,
+
+"""
+
+        return message
+
+    def send_notification(self):
+        """
+        send notification to subscribers
+        """
+
+        snsclient: SNSClient = boto3.client("sns", region_name=self._region_name)
+        # Create topic if not exist
+        for topic in self._conf["sns"]:
+            logger.info("Checking if topic exists")
+            if not self._get_topic(snsclient, topic):
```

Comment:
doesn't the `snsclient.publish` call anyway raise an exception if the topic doesn't exist? so why should we check here?

Reply:
Considering case like having multiple topics, but make sure to send notifications even if only some topics are valid.

Reply:
not sure if I find that better. as written in the review, I would just document that the topics need to exist.
This can also fail if somebody created a FIFO topic and tries to send emails. That doesn't work.